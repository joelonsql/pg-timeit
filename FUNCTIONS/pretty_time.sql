--
-- Returns measured execution_time in human-readable output,
-- using time unit suffixes, i.e. "ns", "us", "ms".
--
CREATE OR REPLACE FUNCTION timeit.pretty_time(numeric)
RETURNS text
LANGUAGE sql
SET search_path = timeit, public, pg_temp
AS $$
SELECT
    CASE
        WHEN $1 IS NULL THEN 'NULL'
        WHEN log10(nullif(abs($1),0)) >= 0 THEN format('%s s', $1)
        WHEN log10(nullif(abs($1),0)) >= -3 THEN format('%s ms', trim_scale($1 * 1e3))
        WHEN log10(nullif(abs($1),0)) >= -6 THEN format('%s Âµs', trim_scale($1 * 1e6))
        WHEN log10(nullif(abs($1),0)) >= -9 THEN format('%s ns', trim_scale($1 * 1e9))
        ELSE format('%s s', $1::float8::text)
    END
$$;

CREATE OR REPLACE FUNCTION timeit.pretty_time(numeric, significant_figures integer)
RETURNS text
LANGUAGE sql
AS $$
SELECT timeit.pretty_time(timeit.round_to_sig_figs($1, $2))
$$;
