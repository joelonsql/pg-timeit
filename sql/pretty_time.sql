SELECT timeit.pretty_time(124999,2);
SELECT timeit.pretty_time(124999,3);
SELECT timeit.pretty_time(0.000124999,2);
SELECT timeit.pretty_time(0.000124999,3);
SELECT timeit.pretty_time(1);
SELECT timeit.pretty_time(0.1);
SELECT timeit.pretty_time(0.01);
SELECT timeit.pretty_time(0.001);
SELECT timeit.pretty_time(0.0001);
SELECT timeit.pretty_time(0.00001);
SELECT timeit.pretty_time(0.000001);
SELECT timeit.pretty_time(0.0000001);
SELECT timeit.pretty_time(0.00000001);
SELECT timeit.pretty_time(0.000000001);
SELECT timeit.pretty_time(0.0000000001);
