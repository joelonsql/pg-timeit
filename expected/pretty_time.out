SELECT timeit.pretty_time(124999,2);
 pretty_time 
-------------
 120000 s
(1 row)

SELECT timeit.pretty_time(124999,3);
 pretty_time 
-------------
 125000 s
(1 row)

SELECT timeit.pretty_time(0.000124999,2);
 pretty_time 
-------------
 120 µs
(1 row)

SELECT timeit.pretty_time(0.000124999,3);
 pretty_time 
-------------
 125 µs
(1 row)

SELECT timeit.pretty_time(1);
 pretty_time 
-------------
 1 s
(1 row)

SELECT timeit.pretty_time(0.1);
 pretty_time 
-------------
 100 ms
(1 row)

SELECT timeit.pretty_time(0.01);
 pretty_time 
-------------
 10 ms
(1 row)

SELECT timeit.pretty_time(0.001);
 pretty_time 
-------------
 1 ms
(1 row)

SELECT timeit.pretty_time(0.0001);
 pretty_time 
-------------
 100 µs
(1 row)

SELECT timeit.pretty_time(0.00001);
 pretty_time 
-------------
 10 µs
(1 row)

SELECT timeit.pretty_time(0.000001);
 pretty_time 
-------------
 1 µs
(1 row)

SELECT timeit.pretty_time(0.0000001);
 pretty_time 
-------------
 100 ns
(1 row)

SELECT timeit.pretty_time(0.00000001);
 pretty_time 
-------------
 10 ns
(1 row)

SELECT timeit.pretty_time(0.000000001);
 pretty_time 
-------------
 1 ns
(1 row)

SELECT timeit.pretty_time(0.0000000001);
 pretty_time 
-------------
 1e-10 s
(1 row)

